---
title: "Xena Workflow with downstream analysis"
author: Maclean Oviatt
output: html_notebook
---


```{r}
suppressMessages(library(UCSCXenaTools))
suppressMessages(library(dplyr))
suppressMessages(library(ggplot2))
suppressMessages(library(survival))
suppressMessages(library(survminer))
suppressMessages(library(dlookr))
suppressMessages(library(TCGAbiolinks))
suppressMessages(library(limma))
suppressMessages(library(tidyr))
```

```{r}
#Obtaining data from TCGA cohorts, based upon name or interested tissue. Prints out a list of corresponding cohorts and datasets to analyze.

TCGA_cohort = XenaData %>%
  filter(XenaHostNames == "tcgaHub") %>% # select TCGA Hub
  XenaScan("TCGA OV")   # select TCGA cohort

TCGA_cohort
```

```{r}
#Based upon the cohort chosen (OV) the clinical dataset is then downloaded and prepared into R as cli. This provides the clinical and survival data for analysis.
cli_query = TCGA_cohort %>%
  filter(DataSubtype == "phenotype") %>%  # select clinical dataset
  XenaGenerate() %>%  # generate a XenaHub object
  XenaQuery() %>%
  XenaDownload()

cli = XenaPrepare(cli_query) #download and prepare 

# See a few rows
head(cli)

#glimpse(cli)
```

```{r}
#Pull all mRNA expression values from TCGA_cohort (NOTE this may vary depending on RNA seq data provided for that specific cohort) Make sure to identify DataSubtype and Label of interest from (TCGA_cohort)
exp_query = TCGA_cohort %>%
   filter(DataSubtype == "gene expression RNAseq", Label == "IlluminaHiSeq UNC") %>%
  XenaGenerate() %>%
  XenaQuery() %>%
  XenaDownload()
expression = XenaPrepare(exp_query)
head(expression)

#transposing the matrix into a dataframe to join Gene expression data with clinical and survival data. 
expression_flip <- t(expression)
colnames(expression_flip) <- expression_flip[1,]
expression_flip <- expression_flip[-c(1),]
expression_flip <- as.data.frame(expression_flip)
```

```{r}
#query data subtype gene expression to pull gene or genes of interest 
ge = TCGA_cohort %>%
  filter(DataSubtype == "gene expression RNAseq", Label == "IlluminaHiSeq UNC")

#Gene or Genes of interest 
gene = "GBP1"

#Pull gene of interest expression values 
Gene = fetch_dense_values(host = ge$XenaHosts,
                          dataset = ge$XenaDatasets,
                          identifiers = gene,
                          use_probeMap = TRUE) %>%
  .[1, ]

#Create dataframe
Gene_express <- tibble(sampleID = names(Gene),
                     Gene_expression = as.numeric(Gene))
as.data.frame(Gene_express)
```

```{r}
#EDA!
describe(Gene_express)
normality(Gene_express)
plot_normality(Gene_express)
head(Gene_express)
summary(Gene_express)
hist(Gene_express$Gene_expression)
ggplot(data=Gene_express, aes(x=Gene_expression)) +
  geom_boxplot(fill="steelblue", color="black") +
  ggtitle("Histogram of GBP1 expression in TCGA Cohort")

hist(Gene_express$Gene_expression, freq = FALSE, xlab = "GBP1 expression", main = "Distribution of expression", col="lightgreen")
curve(dnorm(x, mean = mean(Gene_express$Gene_expression)), add=TRUE, col="darkblue", lwd=3)

eda_web_report(Gene_express)
#It can be observed that although normally distributed the majority of the expression values are in the 9-12 range
```

```{r}
library(data.table)

#Based on EDA seperating high GBP1 expression from low GBP1 expression for upstream analysis. 
"Gene_expression = Gene_express %>%
  mutate(group = case_when(
    Gene_expression > 10.5 ~ '1',
    Gene_expression <= 10.5 ~ '0',
    TRUE ~ NA_character_
  ))
Gene_expression$group <- as.numeric(Gene_expression$group)"

Gene_expression = Gene_express %>%
  mutate(group = case_when(
    Gene_expression > quantile(Gene_expression, 0.8) ~ 'high',
    Gene_expression < quantile(Gene_expression, 0.15) ~ 'low'
  ))

#Function to delete NA's
delete.na <- function(DF, n=0) {
  DF[rowSums(is.na(DF)) <= n,]
}
Gene_expression <- delete.na(Gene_expression)

Gene_expression$group <- as.factor(Gene_expression$group)

#Separation of high and low expression
Gene_low <- Gene_expression[Gene_expression[, "group"]== "low",]
head(Gene_low)
Gene_high <- Gene_expression[Gene_expression[, "group"]== "high",]
head(Gene_high)


hist(Gene_low$Gene_expression)
hist(Gene_high$Gene_expression)

ggplot(Gene_expression, aes(group, fill = factor(group)))+
  geom_bar(position = "fill") +
  geom_text(aes(label = after_stat(count)), stat = "count", position = "fill")
```
```{r}
#transpose matrix for merging with target gene only expression levels 
expression_flip <- as.data.frame(expression_flip)
expression_flip <- tibble::rownames_to_column(expression_flip, "sampleID")


#Expression matrix for TCGA cohort (Used for DGE)
merged_data=
  left_join(Gene_expression, expression_flip, by= "sampleID")
ff <- merged_data[-(1:3)]
merged_data[-(1:3)] <- lapply(ff,as.numeric)
count_matrix <- t(merged_data)
count_matrix <- as.data.frame(count_matrix)
colnames(count_matrix) <- count_matrix[1,]
count_matrix <-count_matrix[-c(1,2,3),]
count_matrix <- count_matrix %>%
  mutate_if(is.character, as.numeric)
count_matrix <- round(count_matrix)

```

#Expression matrix for samples with High GBP1 expression
merged_h = 
  left_join(Gene_high ,expression_flip, by= "sampleID")

#Expression matrix for samples with Low GBP1 expression
merged_l = 
  left_join(Gene_low ,expression_flip, by= "sampleID")

#Count matrix for high and low expression based upon gene target expression
expression_low <- t(merged_l)
colnames(expression_low) <- expression_low[1,]
expression_low <- expression_low[-c(1,2,3),]
expression_low <- as.data.frame(expression_low)
expression_low <- expression_low %>%
  mutate_if(is.character, as.numeric)
expression_low <- tibble::rownames_to_column(expression_low, "sampleID")
expression_low <- as.matrix(expression_low)

write.table(expression_low, "~/low_count_matrix.txt", sep = "\t")

expression_high <- t(merged_h)
colnames(expression_high) <- expression_high[1,]
expression_high <- expression_high[-c(1,2,3),]
expression_high <- as.data.frame(expression_high)
#expression_high <- tibble::rownames_to_column(expression_high, "sampleID")
#expression_high <- as.matrix(expression_high)

write.table(expression_high, "~/high_count_matrix.txt", sep = "\t")


```{r}
#Renaming of sampleID in order to merge datasets
cli$OV_survival.txt = rename(cli$OV_survival.txt, sampleID = sample)

#Clinical data with Gene expression
cli_data =
  left_join(Gene_expression,cli$OV_clinicalMatrix, by= "sampleID") 
head(cli_data)

#Survival data with Gene expression
survival_df =
  left_join(merged_data,cli$OV_survival.txt, by="sampleID") %>%
  dplyr::select(sampleID, Gene_expression, group, OS.time, OS,) %>%
  rename(time = OS.time, status = OS)
head(survival_df)

#GBP1 expression with respect to Cancer Grade
ggplot(data = cli_data) +
  geom_histogram(mapping = aes(x = Gene_expression), binwidth = 0.1)+
  facet_wrap(~neoplasm_histologic_grade)
```
#Survival data with GBP1 expression matrix
survival_gene =
  left_join(Gene_expression, cli$OV_survival.txt, by="sampleID") %>%
  select(sampleID, GBP1_expression, OS.time, OS,) %>%
  rename(time = OS.time, status = OS)
head(survival_GBP1)

```{r}
#EDA of clinical and survival data with GBP1_expression

# install.packages("SmartEDA")
library(SmartEDA)

# similarly, with dplyr syntax: df %>% ExpReport(...)
ExpReport(
  cli_data,
  Target="Gene_expression",
  label=NULL,
  op_file="Report.html",
  op_dir=getwd())

#EDA of GBP1 expression with clinical data
ExpData(data=cli_data, type=2)
ExpCatStat(cli_data, Target = "Gene_expression", Pclass= "1")

library(DataExplorer)
cli_data %>%
    create_report(
        output_file = "~/Target_gene_clinical.html",
        report_title = "EDA Report - Target gene clinical Dataset",
        y = "Gene_expression"
    )

survival_df %>%
  create_report(
    output_file = "~/Target_gene_survival.html",
    report_title = "EDA_Report -Target Gene Survival Dataset",
    y = "Gene_expression"
  )

```

```{r}

#PLOTS (GBP1 high expression with clinical phenotypes investigation)
ggplot(Gene_high_cli, aes(neoplasm_histologic_grade))+ geom_bar()
ggplot(Gene_high_cli, aes(lymphatic_invasion)) + geom_bar()
```



```{r}
library(biomaRt)
library(PCAtools)
library(EnhancedVolcano)

## Discard genes, which are expressed in less than 20% of all samples (considering that we have 2 conditions in total )

##  % selection is based on no. of conditions  ## more condition means less %
keep <- rowSums(count_matrix > 0) >= round(ncol(count_matrix)*.20)  
count_matrix <- count_matrix[keep,]


##add rownames as sample name (if not already), because pca function check rownames of anno == col of data matrix
## sort anno based on condition ## good representation in heatmap 
anno <- Gene_expression 
anno <- anno[,-2]

rownames(anno) <- anno$sampleID

## log transformation of rawcount for PCA plot 
lograwcount <- as.matrix(log2(count_matrix +1))

## taking top 1000 genes having highest variance selected from all the genes in the input
top1000.order <- head(order(matrixStats::rowVars(lograwcount), decreasing = TRUE), 1000) 

pp <- PCAtools::pca(lograwcount[top1000.order,], metadata =anno, removeVar = 0.01) ## performing PCA

 #visualization of PCA plot
biplot(pp)
  
#Create DESeq2 datasets 

dds <- DESeqDataSetFromMatrix(countData = count_matrix, colData = anno, design = ~group)

#View(counts(dds))

dds <- estimateSizeFactors(dds)

 ### Transform counts for data visualization
vst <- vst(dds, blind=TRUE) 

#Normalize the data
normalized_counts <- as.data.frame(assay(vst))

#Plot PCA
plotPCA(vst, intgroup="group") 

# Run DESeq2 (Differential Gene Expression Analysis)
dds <- DESeq(dds, fitType='local')

##ensure your data is a good fit for the DESeq2 model
plotDispEsts(dds)

#Extract result dateframe
res <- results(dds)
View(as.data.frame(res))

# Filter differentially expressed genes
alpha <- 0.05
log2FoldChangeThreshold <- 1
sigGenes <- subset(res, padj < alpha & abs(log2FoldChange) > log2FoldChangeThreshold)

# Plot the results
plotMA(res, alpha = alpha, ylim = c(-10, 10))

# Heatmap of differentially expressed genes
heatmap(assay(dds)[rownames(sigGenes), ], scale = "row", show_rownames = FALSE, col = colorRampPalette(c("blue", "white", "red"))(100))

# Generate a normalized count matrix
normCountMatrix <- counts(dds, normalized = TRUE)
```

```{r}
#Volcano Plot

library(EnhancedVolcano)

volcano <- EnhancedVolcano(res,
                lab = rownames(res),
                x = 'log2FoldChange',
                pCutoff = 1e-05,
                FCcutoff = 1,
                y = 'pvalue')
volcano
```

```{r}

#Threshold definition for cutoff
p.threshold <- 0.05
res$threshold <- as.logical(res$padj < p.threshold)  

#Up regulated vs. Down regulated genes 
#adding extra information, in which condition it's down or up 
firstC <- "High"
SecondC <- "Low"
nam <- paste('down_in',firstC,SecondC, sep = '_')   
res[, nam] <- as.logical(res$log2FoldChange < 0)  


 ### list of gene with Padjust < defined threshold
genes.deseq <- row.names(res)[which(res$threshold)]  
genes_deseq2_sig <- as.data.frame(res[which(res$threshold),])
View(as.data.frame(genes_deseq2_sig))

genes_deseq2_sig$gene <- row.names(genes_deseq2_sig)

#Thresholds for Up/Down regulation 
foldChangeThreshold <- 2  # Adjust as needed
pValueThreshold <- 0.05   # Adjust as needed

# Identify up-regulated significant genes
upregulatedGenes <- rownames(genes_deseq2_sig)[genes_deseq2_sig$log2FoldChange > log2(foldChangeThreshold) & genes_deseq2_sig$padj < pValueThreshold] 

# Identify down-regulated significant genes
downregulatedGenes <- rownames(genes_deseq2_sig)[genes_deseq2_sig$log2FoldChange < -log2(foldChangeThreshold) & genes_deseq2_sig$padj < pValueThreshold]


cat("Number of upregulated genes:", length(upregulatedGenes), "\n")
cat("Number of downregulated genes:", length(downregulatedGenes), "\n")

# Save the lists of upregulated and downregulated genes
write.table(upregulatedGenes, "~/upregulated_genes.txt", quote = FALSE, row.names = FALSE)
write.table(downregulatedGenes, "~/downregulated_genes.txt", quote = FALSE, row.names = FALSE)
```

```{r}
########### Plots normalized count of top 20 genes ## sorted based on padjust and filter by |logFC| >=1

res$gene <- row.names(res)
View(as.data.frame(res))

top20 <- res %>% 
  as.data.frame %>%
  arrange(padj) %>% 	#Arrange rows by padj values
  filter(abs(log2FoldChange) >=1.5) %>%   #filter based on logFC
  pull(gene) %>% 		#Extract character vector of ordered genes
  head(n=20) 		#Extract the first 20 genes

top20_norm <- as.data.frame(normalized_counts[rownames(normalized_counts) %in% top20,])

top20_norm_v2 <- top20_norm ## will use later for heatmap

top20_norm <- (top20_norm+1) ## in later step to remove infinity bias due to log

top20_norm$gene <-  row.names(top20_norm)

top20_norm <- top20_norm %>% 
  pivot_longer(!gene, names_to = "samplename", values_to = "normalized_counts") # Gathering the columns to have normalized counts to a single column

# Create tibbles including row names #for anno
mov10_meta <- anno %>% 
  rownames_to_column(var="samplename") %>% 
  as_tibble()

top20_norm <- inner_join(mov10_meta, top20_norm)  ## we are merging anno of 20 with previously extract format to draw dot plot

################
## plot using ggplot2

library(ggplot2)

ggplot(top20_norm) +
  geom_point(aes(x = gene, y = normalized_counts, color = group)) +
  scale_y_log10() + 
  xlab("Genes") +
  ylab("log 10 CPM Normalized Counts") +
  ggtitle("Top 20 Significant DE Genes with abs(logFC) =>1") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(plot.title = element_text(hjust = 0.5))


file <- paste('Deseq2_',firstC,'_v_',SecondC,'_results_significant_padj',p.threshold,'.csv',sep = '') 
all_results <- paste('Deseq2_',firstC,'_v_',SecondC,'_all_results.csv',sep = '')

res <- as.data.frame(res)
View(res)

write.table(genes_deseq2_sig, file, sep = ",") ## significant results (including thresholds)
write.table(res,all_results,sep = ",")  ## no LogFC threshold (all results)

```
```{r}
### Option (1) continue from DGE analysis or (2) upload data from saved file of DEG 
#genes_deseq2_sig <- read.csv("Deseq2_high_v_low_results_significant_padj0.05.csv") 

###################### Extract various types of gene ids from Biomart ########################
library(biomaRt)

### define the mart for h_sapiens

ensembl_mart <- useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")  ## either this or following line
#ensembl_mart <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "asia") ## takes little bit time

###### following lines is extarcting other alternate names of the hugo gene symobols 
###### in this case entrez gene ids
sig_genes <- as.data.frame(genes_deseq2_sig)
genes.entrezid <- getBM(attributes = c("hgnc_symbol", "entrezgene_id"), filters = "hgnc_symbol", values = sig_genes$gene, mart = ensembl_mart)

#genes.entrezid = as.data.frame(genes.entrezid) ## if not defined as df

merged <- merge(sig_genes, genes.entrezid, by.x= "gene", by.y="hgnc_symbol")


######### Rank all genes based on their fold change #########

#BiocManager::install("clusterProfiler", force = TRUE)
#BiocManager::install("pathview", force = TRUE)
#BiocManager::install("enrichplot", force = TRUE)

library(clusterProfiler)
library(enrichplot)
library(ggplot2)

# SET THE DESIRED ORGANISM HERE ###https://learn.gencore.bio.nyu.edu/rna-seq-analysis/gene-set-enrichment-analysis/
organism = "org.Hs.eg.db"  ## search other organism annotations here http://bioconductor.org/packages/release/BiocViews.html#___OrgDb

#BiocManager::install(organism, character.only = TRUE, force = TRUE)

library(organism, character.only = TRUE)
#We will take the log2FoldChange value from previously saved significant results file
#Deseq2_case1_v_Control_results_significant.csv


# we want the log2 fold change 
original_gene_list <- merged$log2FoldChange
#print(original_gene_list)

# name the vector
names(original_gene_list) <- merged$entrezgene_id
#print(original_gene_list)

# omit any NA values  ## excluding genes where no entrez ids available ## you may loose some information 
gene_list1<-na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list1 = sort(gene_list1, decreasing = TRUE)

View(as.data.frame(gene_list1))
#### Gene Set Enrichment  of Gene Ontology #####

library(stats)

keytypes(org.Hs.eg.db)

 #### Try GO with all different ont methods parameter  ## BP = Biological Processes, CC= Cellular component, MF = Molecular functions
gse <- gseGO(geneList=gene_list1, 
             ont ='ALL', 
             keyType = "ENTREZID", 
             nPerm = 1000,
             minGSSize = 20, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = org.Hs.eg.db, 
             pAdjustMethod = "none")

 require(DOSE)
 view(as.data.frame(gse))
 dotplot(gse, showCategory=10, split=".sign", orderBy = "X") + facet_grid(.~.sign)
 gseaplot(gse, by = "all", title = gse$Description[1], geneSetID = 1)
 ridgeplot(gse) + labs(x = "enrichment distribution")
```
```{r}
 


##GO Enrichment Analysis of a gene set. 
##Given a vector of genes, enrichGO function will return the 
##enrichment GO categories after FDR control.


```

```{r}
#library(clusterProfiler)
#library(org.Hs.eg.db)
library(enrichplot)
library(GOSemSim)
library(ggnewscale)
library(DOSE)

# Exctract significant results (padj < 0.05)
sig_genes_df = subset(merged, padj < 0.05)

# From significant results, we want to filter on log2fold change
genes <- sig_genes_df$log2FoldChange

# Name the vector
names(genes) <- sig_genes_df$entrezgene_id

# omit NA values
genes <- na.omit(genes)

# filter on min log2fold change (log2FoldChange > 2)
genes <- names(genes)[abs(genes) > 2]

ego <- enrichGO(gene  = genes,
                OrgDb = org.Hs.eg.db,
                ont = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.10,
                readable= T)
ego_res <- as.data.frame(ego@result)
d <- godata('org.Hs.eg.db', ont="BP") #prepare GO DATA for measuring semantic similarity ## required for next step

ego2 <- pairwise_termsim(ego, method="Wang", semData = d) #enrichment result #method of calculating the similarity between nodes #GOSemSimDATA object
```

