---
title: "Xena_Workflow"
author: Maclean Oviatt
output: html_notebook
---


```{r}
suppressMessages(library(UCSCXenaTools))
suppressMessages(library(dplyr))
suppressMessages(library(ggplot2))
suppressMessages(library(survival))
suppressMessages(library(survminer))
suppressMessages(library(dlookr))
suppressMessages(library(TCGAbiolinks))
suppressMessages(library(limma))
library(tidyr)

```


```{r}
#Obtaining data from TCGA cohorts, based upon name or interested tissue. Prints out a list of corresponding cohorts and datasets to analyze.

TCGA_cohort = XenaData %>%
  filter(XenaHostNames == "tcgaHub") %>% # select TCGA Hub
  XenaScan("TCGA OV")   # select TCGA cohort

TCGA_cohort
```
```{r}
#PATHWAYS?
"path_query = TCGA_cohort %>%
   filter(DataSubtype == "pathway activity", Label == "Paradigm IPLs") %>%
  XenaGenerate() %>%
  XenaQuery() %>%
  XenaDownload()
pathways = XenaPrepare(path_query)"
```


```{r}
#Based upon the cohort chosen (OV) the clinical dataset is then downloaded and prepared into R as cli. This provides the clinical and survival data for analysis.
cli_query = TCGA_cohort %>%
  filter(DataSubtype == "phenotype") %>%  # select clinical dataset
  XenaGenerate() %>%  # generate a XenaHub object
  XenaQuery() %>%
  XenaDownload()

cli = XenaPrepare(cli_query) #download and prepare 

# See a few rows
head(cli)

#glimpse(cli)
```

```{r}
#Pull all mRNA expression values from TCGA_cohort (NOTE this may vary depending on RNA seq data provided for that specific cohort) Make sure to identify DataSubtype and Label of interest from (TCGA_cohort)
exp_query = TCGA_cohort %>%
   filter(DataSubtype == "gene expression RNAseq", Label == "IlluminaHiSeq UNC") %>%
  XenaGenerate() %>%
  XenaQuery() %>%
  XenaDownload()
expression = XenaPrepare(exp_query)
head(expression)

#transposing the matrix into a dataframe to join GBP1 expression data with clinical and survival data. 
expression_flip <- t(expression)
colnames(expression_flip) <- expression_flip[1,]
expression_flip <- expression_flip[-c(1),]
expression_flip <- as.data.frame(expression_flip)

```

```{r}
#query data subtype gene expression to pull gene or genes of interest 
ge = TCGA_cohort %>%
  filter(DataSubtype == "gene expression RNAseq", Label == "IlluminaHiSeq UNC")

#Gene or Genes of interest 
gene = "GBP1"

#Pull gene of interest expression values 
GBP1 = fetch_dense_values(host = ge$XenaHosts,
                          dataset = ge$XenaDatasets,
                          identifiers = gene,
                          use_probeMap = TRUE) %>%
  .[1, ]

#Create dataframe
GBP1_OV <- tibble(sampleID = names(GBP1),
                     GBP1_expression = as.numeric(GBP1))
as.data.frame(GBP1_OV)
```

```{r}
#EDA!
describe(GBP1_OV)
normality(GBP1_OV)
plot_normality(GBP1_OV)
eda_report(GBP1_OV)
head(GBP1_OV)
summary(GBP1_OV)
hist(GBP1_OV$GBP1_expression)
ggplot(data=GBP1_OV, aes(x=GBP1_expression)) +
  geom_boxplot(fill="steelblue", color="black") +
  ggtitle("Histogram of GBP1 expression in TCGA Cohort")

hist(GBP1_OV$GBP1_expression, freq = FALSE, xlab = "GBP1 expression", main = "Distribution of expression", col="lightgreen")
curve(dnorm(x, mean = mean(GBP1_OV$GBP1_expression)), add=TRUE, col="darkblue", lwd=3)

#It can be observed that although normally distributed the majority of the expression values are in the 9-12 range
```

```{r}

library(data.table)
#Based on EDA seperating high GBP1 expression from low GBP1 expression for upstream analysis. 
GBP1_OV = GBP1_OV %>%
  mutate(group = case_when(
    GBP1_expression => 10.5 ~ '1',
    GBP1_expression <= 10.5 ~ '0',
    TRUE ~ NA_character_
  ))
GBP1_OV$group <- as.numeric(GBP1_OV$group)

GBP1_OV = GBP1_OV %>%
  mutate(group = case_when(
    GBP1_expression > quantile(GBP1_expression, 0.9) ~ '1',
    GBP1_expression < quantile(GBP1_expression, 0.1) ~ '0'
  ))
na.omit(GBP1_OV)
GBP1_OV %>%
  drop_na(.)

GBP1_low <- GBP1_OV[GBP1_OV[, "group"]== 0,]
head(GBP1_low)
GBP1_high <- GBP1_OV[GBP1_OV[, "group"]== 1,]
head(GBP1_high)

hist(GBP1_low$GBP1_expression)
hist(GBP1_high$GBP1_expression)

ggplot(GBP1_OV, aes(group, fill = factor(group)))+
  geom_bar(position = "fill") +
  geom_text(aes(label = after_stat(count)), stat = "count", position = "fill")

#names(pathways)[names(pathways) == 'sample'] <- 'sampleID'

#path_GBP1_high = 
  #left_join(GBP1_OV, pathways, by= "sampleID")

#path_GBP1_low =
  #left_join(pathways, GBP1_low, by= 'sampleID')

GBP1_OV$group <- as.numeric(GBP1_OV$group) %>%
  GBP1_expression > quantile(GBP1_expression, 0.5) ~ '1',
GBP1_expression < quantile(GBP1_expression, 0.5) ~ '0',
    TRUE ~ NA_character_
  ))
#summary(GBP1)"

```

```{r}
#EDA of clinical and survival data with GBP1_expression

# install.packages("SmartEDA")
library(SmartEDA)

# similarly, with dplyr syntax: df %>% ExpReport(...)
ExpReport(
  cli_data,
  Target="GBP1_expression",
  label=NULL,
  op_file="Report.html",
  op_dir=getwd())

#EDA of GBP1 expression with clinical data
ExpData(data=cli_data, type=2)
ExpCatStat(cli_data, Target = "GBP1_expression", Pclass= "1")

library(DataExplorer)
cli_data %>%
    create_report(
        output_file = "~/GBP1_clinical.html",
        report_title = "EDA Report - Cardiovascular Disease Dataset",
        y = "GBP1_expression"
    )

survival_GBP1 %>%
  create_report(
    output_file = "~/GBP1_survival.html",
    report_title = "EDA_Report -GBP1 expression Survival Dataset",
    y = "GBP1_expression"
  )


plot_str(cli_data)
plot_bar(cli_data)
```


```{r}
#transpose matrix for merging with GBP1 only expression levels 
expression_flip <- as.data.frame(expression_flip)
expression_flip <- tibble::rownames_to_column(expression_flip, "sampleID")

#Expression matrix for TCGA cohort
merged_data=
  left_join(GBP1_OV, expression_flip, by= "sampleID")
ff <- merged_data[,c(3:20533)]
merged_data[,c(3:20533)] <- lapply(ff,as.numeric)

#Expression matrix for samples with High GBP1 expression
merged_h = 
  left_join(GBP1_high ,expression_flip, by= "sampleID")

#Expression matrix for samples with Low GBP1 expression
merged_l = 
  left_join(GBP1_low ,expression_flip, by= "sampleID")
```

```{r}
#Transpose expression matrix for DGE analysis
expression_low <- t(merged_l)
colnames(expression_low) <- expression_low[1,]
expression_low <- expression_low[-c(1,2),]
expression_low <- as.data.frame(expression_low)
expression_low <- expression_low %>%
  mutate_if(is.character, as.numeric)
expression_low <- tibble::rownames_to_column(expression_low, "sample")
expression_low <- as.matrix(expression_low)

merged_data$group <- as.factor(merged_data$group)
merged_try <- t(merged_data)
colnames(merged_try) <- merged_try[1,]
merged_try <- merged_try[-c(1,2,3),]
merged_try <- as.data.frame(merged_try)
ff <- merged_try[,c(1:308)]
merged_try[,c(1:308)] <- lapply(ff,as.numeric)


expression_high <- t(merged_h)
colnames(expression_high) <- expression_high[1,]
expression_high <- expression_high[-c(1,3),]
expression_high <- as.data.frame(expression_high)
expression_high <- tibble::rownames_to_column(expression_high, "sample")
expression_high <- as.matrix(expression_high)
```

```{r}
#Renaming of sampleID in order to merge datasets
cli$OV_survival.txt = rename(cli$OV_survival.txt, sampleID = sample)

#Clinical data with GBP1 expression
cli_data =
  left_join(GBP1_OV,cli$OV_clinicalMatrix, by= "sampleID") 
head(cli_data)

#Survival data with expression matrix
survival_df =
  left_join(merged_data,cli$OV_survival.txt, by="sampleID") %>%
  #select(sampleID, GBP1_expression, group, OS.time, OS,) %>%
  rename(time = OS.time, status = OS)
head(survival_df)

#Survival data with GBP1 expression matrix
survival_GBP1 =
  left_join(GBP1_OV, cli$OV_survival.txt, by="sampleID") %>%
  select(sampleID, GBP1_expression, OS.time, OS,) %>%
  rename(time = OS.time, status = OS)
head(survival_GBP1)

#GBP1 expression with respect to Cancer Grade
ggplot(data = cli_data) +
  geom_histogram(mapping = aes(x = GBP1_expression), binwidth = 0.1)+
  facet_wrap(~neoplasm_histologic_grade)
```


```{r}
glimpse(cli_data)
cli_data$group <- as.factor(cli_data$group)
survival_df$group <- as.factor(survival_df$group)

#Subsetting for high GBP1 expression
GBP1_high_cli <- cli_data[cli_data[, "group"]== "1",]
head(GBP1_high_cli)

GBP1_High_ex <- merged_data[merged_data[, "group"]=="GBP1_High",]
ff <- GBP1_High_ex[,c(4:20533)]
GBP1_High_ex[,c(4:20533)] <- lapply(ff,as.numeric)
head(GBP1_High_ex)

#PLOTS (GBP1 high expression with clinical phenotypes investigation)
ggplot(GBP1_high_cli, aes(neoplasm_histologic_grade))+ geom_bar()
ggplot(GBP1_high_cli, aes(lymphatic_invasion)) + geom_bar()
```

```{r}

cat <- lm(as.matrix(merged_data)~ merged_data$group)

dge <- DGEList(counts = merged_try, group = merged_data$group)

design <- as.matrix(merged_h)

dge <- calcNormFactors(dge)

logCPM<- cpm(dge, log = TRUE, prior.count = 3)

fit <- lmFit(logCPM)
fit <- eBayes(fit)
topTable(fit)
dim(fit)
fit2 <- treat(fit, lfc= 0.1)

fit <- eBayes()
fit <- topT
```


```{r}
fit = coxph(Surv(time, status) ~ GBP1_expression , data = survival_df)
fit
```
```{r}

fit = survfit(Surv(time, status) ~ group, data = survival_df)
ggsurvplot(fit, pval= TRUE)
```

